### PR TITLE
AirfRANS: lr=7e-4 + grad-clip=1.0 + WD=1e-2 — combine grad-clip + regularization

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

A previous WD=1e-2 experiment failed badly (0.027) because it lacked gradient clipping — the instability from high LR at cosine peaks overwhelmed any regularization benefit. Now that grad-clip=1.0 has been proven to stabilize training at lr=7e-4 (PR #2680, 0.01419), WD=1e-2 can be layered on top from a stable foundation. If WD=1e-2 adds beneficial regularization (reducing overfitting to training distribution), the combination should be strictly better than grad-clip=1.0 alone. The previous WD failure was a training stability problem, not a capacity problem — grad-clip fixes the stability, so WD should now contribute positively.

## Instructions

Run the same configuration as PR #2680 (lr=7e-4, T_max=10, grad-clip=1.0, Fourier, no-EMA), but add `--weight-decay 1e-2`. The AdamW optimizer already supports weight decay — just add the flag.

Run with `--epochs 999` and let the 30-min timeout govern duration. Report:
- Best val `surface_mse` and at which epoch it occurred
- Whether training is stable (no divergence or extreme spikes)
- W&B run ID and link

## Baseline

- **AirfRANS current best:** 0.01419 val / 0.01513 test (PR #2680, lr=7e-4+grad-clip=1.0, T_max=10, 41 epochs, 30-min)
- **Previous WD=1e-2 result (no grad-clip):** 0.027 — failed due to instability
- **TandemFoil current best:** 52.81 (PR #2595, 5L/256d, Lion lr=3e-4, T_max=10)
- **DrivAerML current best:** 5.73% (PR #2602, 4L/384d/6H, T_max=30, 151 epochs)
- **Baseline W&B run:** 48ldl625 (PR #2680, grad-clip=1.0, no WD)

## Reproduce command

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier --epochs 999 \
  --wandb-group airfrans-gradclip-compounds \
  --wandb-name "fern/airfrans-lr7e4-gradclip-wd1e2"
```

> **Note:** `--epochs 999` and `--no-use-ema` are both mandatory. Default epochs=2 will not train long enough. EMA has a known bug.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Primary Metric — NEW BEST

| Metric | Baseline (grad-clip only) | This Run (grad-clip + WD=1e-2) | Delta |
|--------|--------------------------|-------------------------------|-------|
| val_primary/surface_mse | 0.01419 | **0.01323** (epoch 41) | **-6.8%** |
| test_primary/surface_mse | 0.01513 | **0.01478** (epoch 41) | **-2.3%** |

**Both val and test beat the baseline.** Test was evaluated at the best val epoch (41), so the metric is representative.

### Full Metrics (Epoch 41 — best val AND final epoch)

| Metric | Value |
|--------|-------|
| val_primary/surface_mse | **0.01323** |
| full_val/surface_mse_p | 0.0528 |
| full_val/surface_mse_Ux | 1.15e-4 |
| full_val/surface_mse_Uy | 1.93e-5 |
| full_val/surface_mse_nut | 7.72e-6 |
| full_val/volume_mse | 0.0804 |
| test_primary/surface_mse | **0.01478** |
| full_test/surface_mse_p | 0.0590 |
| full_test/volume_mse | 0.0894 |

### Phase Transition Trajectory

| Epoch | val_primary | surface_mse_p | Phase |
|-------|------------|---------------|-------|
| 1 | 0.3309 | 1.321 | Early |
| 8 | 0.2192 | 0.876 | Plateau descent |
| 12 | 0.1239 | 0.494 | Transition onset |
| **14** | **0.0548** | **0.219** | **First deep transition** |
| 24 | 0.0343 | 0.137 | Second wave |
| 30 | 0.0299 | 0.119 | Deepening |
| 34 | 0.0194 | 0.077 | Near-target |
| 39 | 0.0183 | 0.073 | Sub-baseline |
| **41** | **0.0132** | **0.053** | **New record** |

### Volatility Analysis (vs baseline grad-clip=1.0 only)

| Metric | Baseline (wd=1e-4) | This run (wd=1e-2) |
|--------|--------------------|--------------------|
| Phase transition onset | ~epoch 14 | epoch 14 (same) |
| 7-epoch new-best streak | Not observed | Epochs 6-12 |
| Post-transition oscillation amplitude | Moderate | Similar |
| Final convergence | 0.01419 | **0.01323** |
| Still improving at cutoff | Unclear | **Yes — final epoch was best** |

### Experiment Details

- **Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --epochs 999 --agent fern --wandb-name "fern/airfrans-lr7e4-gradclip-wd1e2" --wandb-group airfrans-gradclip-compounds`
- **W&B Run:** `7vic8kxn`
- **Epochs completed:** 41 (hit 30-min timeout)

### What happened

**The hypothesis is validated — grad-clip + WD=1e-2 beats grad-clip alone.**

1. **New AirfRANS record:** val=0.01323, test=0.01478 — both below the grad-clip-only baseline (0.01419/0.01513). The combination of gradient clipping (stability) and strong weight decay (regularization) produces the best result seen on AirfRANS.

2. **Smooth early training:** 7 consecutive new-best epochs (6-12) before the phase transition, indicating the WD provides beneficial regularization during the plateau phase.

3. **Phase transition at epoch 14:** A dramatic drop from 0.1239 to 0.0548 — similar timing to the grad-clip-only baseline, confirming the transition is driven by the optimizer dynamics, not the WD.

4. **Critical: best at final epoch (41):** The model was STILL improving at cutoff — the last epoch was the best. This strongly suggests a longer training budget would push even deeper. The grad-clip-only baseline peaked at epoch 35 and then oscillated, whereas this run kept descending.

5. **Previous WD failure explained:** The earlier WD=1e-2 run without grad-clip (0.027) failed because the high LR at cosine peaks created gradient explosions that WD alone couldn't control. Grad-clip prevents the extreme gradients, and THEN WD can contribute positive regularization.

### Suggested follow-ups

1. **Longer training budget (60-90 min):** The model was best at epoch 41 (the cutoff) — there's clear headroom. This is the highest-priority follow-up.
2. **wd=5e-3 with grad-clip=1.0:** Test if less aggressive WD gives even better results by allowing slightly deeper basin access while still regularizing.
3. **This is a potential new golden config for AirfRANS:** lr=7e-4 + T_max=10 + grad-clip=1.0 + wd=1e-2 + Fourier + no-EMA.

**Comment by tcapelle** [comment]
ADVISOR: Merging — **0.01323** val_primary/surface_mse is a 6.8% improvement over the 0.01419 baseline. Grad-clip + WD=1e-2 compounds beautifully: clipping prevents gradient explosions so WD can contribute positive regularization. The fact that the best epoch was the FINAL epoch (41) is very promising — more training time should push even deeper. This is now the golden config for AirfRANS: lr=7e-4+T_max=10+grad-clip=1.0+WD=1e-2. Excellent discovery fern!
